Deriving BoundedArbitrary fails with generic type defaults

<!--
If this is a security issue, please report it following the
[security reporting procedure](https://github.com/model-checking/kani/security/policy).
-->

<!--
Thank you for filing a bug report! 🐛
Please provide a short summary of the issue, along with the information necessary to replicate.
-->

I tried this code:
<!--
If the reproducing code is small, please post it here.
Otherwise, please attach the relevant files to this issue.
Alternatively, post a link to the repository and branch that exposes the issue.
-->
```rust
#[derive(kani::BoundedArbitrary)]
pub struct Foo<T = i32> {
    #[bounded]
    v: Vec<T>,
}

#[kani::proof]
fn main() {
    let f: Foo<u8> = kani::bounded_any::<_, 16>();
    assert!(f.v.len() <= 16);
}
```

using the following command line invocation:

```
kani ba.rs
```

with Kani version: a0f936434ca

I expected to see this happen: Verification succeeds

<!--
If Kani crashed, please include a backtrace in the code block by setting
`RUST_BACKTRACE=1` in your environment. E.g. `RUST_BACKTRACE=1 kani ...`
-->
Instead, this happened: 
```bash
$ kani ba.rs 
Kani Rust Verifier 0.62.0 (standalone)
error: expected one of `!`, `(`, `+`, `,`, `::`, `<`, or `{`, found `:`
 --> ba.rs:1:10
  |
1 | #[derive(kani::BoundedArbitrary)]
  |          ^^^^^^^^^^^^^^^^^^^^^^ expected one of 7 possible tokens
  |
  = note: this error originates in the derive macro `kani::BoundedArbitrary` (in Nightly builds, run with -Z macro-backtrace for more info)

error: proc-macro derive produced unparsable tokens
 --> ba.rs:1:10
  |
1 | #[derive(kani::BoundedArbitrary)]
  |          ^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `Foo<u8>: kani::BoundedArbitrary` is not satisfied
  --> ba.rs:9:42
   |
9  |     let f: Foo<u8> = kani::bounded_any::<_, 16>();
   |                                          ^ the trait `kani::BoundedArbitrary` is not implemented for `Foo<u8>`
   |
   = help: the following other types implement trait `kani::BoundedArbitrary`:
             std::collections::HashMap<K, V, std::hash::BuildHasherDefault<std::hash::DefaultHasher>>
             std::collections::HashSet<V, std::hash::BuildHasherDefault<std::hash::DefaultHasher>>
             std::option::Option<T>
             std::result::Result<T, E>
             std::string::String
             std::vec::Vec<T>
note: required by a bound in `kani::bounded_any`
  --> /home/ubuntu/git/kani/library/kani/src/lib.rs:55:1
   |
55 | kani_core::kani_lib!(kani);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `bounded_any`
   = note: this error originates in the macro `kani_core::kani_intrinsics` which comes from the expansion of the macro `kani_core::kani_lib` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 3 previous errors
```

--- a/library/kani_macros/src/derive_bounded.rs
+++ b/library/kani_macros/src/derive_bounded.rs
@@ -53,13 +53,14 @@ pub(crate) fn expand_derive_bounded_arbitrary(
     };
 
     // add `T: Arbitrary` bounds for generics
-    let (generics, clauses) = quote_generics(&parsed.generics);
+    let clauses = quote_generics(&parsed.generics);
+    let (impl_generics, ty_generics, _) = parsed.generics.split_for_impl();
     let name = &parsed.ident;
 
     // generate the implementation
     let kani_path = kani_path();
     quote! {
-        impl #generics #kani_path::BoundedArbitrary for #name #generics
+        impl #impl_generics #kani_path::BoundedArbitrary for #name #ty_generics
             #clauses
         {
             fn bounded_any<const N: usize>() -> Self {
@@ -127,12 +128,14 @@ fn union_constructor(ident: &syn::Ident, _data_union: &syn::DataUnion) -> TokenS
 ///     ...
 /// }
 /// ```
-fn quote_generics(generics: &syn::Generics) -> (TokenStream, TokenStream) {
+fn quote_generics(generics: &syn::Generics) -> TokenStream {
     let kani_path = kani_path();
-    let params = generics.type_params().map(|param| quote!(#param)).collect::<Vec<_>>();
-    let where_clauses = generics.type_params().map(|param| quote!(#param : #kani_path::Arbitrary));
-    if !params.is_empty() {
-        (quote!(<#(#params),*>), quote!(where #(#where_clauses),*))
+    let where_clauses = generics.type_params().map(|param| {
+        let ident = &param.ident;
+        quote!(#ident : #kani_path::Arbitrary)
+    });
+    if generics.type_params().count() > 0 {
+        quote!(where #(#where_clauses),*)
     } else {
         Default::default()
     }

--- a/tests/expected/derive-bounded-arbitrary/generic_default.expected
+++ b/tests/expected/derive-bounded-arbitrary/generic_default.expected
@@ -1,0 +1,5 @@
+Checking harness check_my_vec...
+
+ ** 2 of 2 cover properties satisfied
+
+VERIFICATION:- SUCCESSFUL

--- a/tests/expected/derive-bounded-arbitrary/generic_default.rs
+++ b/tests/expected/derive-bounded-arbitrary/generic_default.rs
@@ -1,0 +1,24 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Check that derive BoundedArbitrary macro works on structs with a generic default
+//! which had an issue in the past:
+//! https://github.com/model-checking/kani/issues/4116
+
+extern crate kani;
+use kani::BoundedArbitrary;
+
+#[derive(BoundedArbitrary)]
+#[allow(unused)]
+struct MyVector<T = i32> {
+    #[bounded]
+    vector: Vec<T>,
+}
+
+#[kani::proof]
+#[kani::unwind(6)]
+fn check_my_vec() {
+    let my_vec: MyVector<u8> = kani::bounded_any::<_, 1>();
+    kani::cover!(my_vec.vector.len() == 0);
+    kani::cover!(my_vec.vector.len() == 1);
+}

